repos/system_upgrade/supplements/actors/checkreboothygiene/libraries/checkreboothygiene.py

```diff
@@ -0,0 +1,126 @@
+import os
+import time
+
+from leapp import reporting
+from leapp.exceptions import StopActorExecutionError
+from leapp.libraries.stdlib import CalledProcessError, api, run
+
+DAY_IN_SECONDS = 24 * 60 * 60
+# TODO: can this be parametrized? (Makefile)
+EXCESSIVE_UPTIME_LIMIT_DAYS = 30
+EXCESSIVE_UPTIME_LIMIT_SECONDS = EXCESSIVE_UPTIME_LIMIT_DAYS * DAY_IN_SECONDS
+
+
+def check_excessive_uptime():
+    uptime_seconds = _get_uptime()
+
+    if uptime_seconds > EXCESSIVE_UPTIME_LIMIT_SECONDS:
+        summary = (
+            "This host has not been rebooted for over {} days. "
+            "To reduce the risk of boot issues related to any changes made since the last reboot, "
+            "policy requires the host to be rebooted before going forward with the upgrade."
+        ).format(EXCESSIVE_UPTIME_LIMIT_DAYS)
+
+        report = [
+            reporting.Title("Excessive uptime detected"),
+            reporting.Summary(summary),
+            reporting.Severity(reporting.Severity.HIGH),
+            reporting.Groups([reporting.Groups.SANITY, reporting.Groups.INHIBITOR]),
+            reporting.Remediation(hint="Please reboot the host machine."),
+        ]
+        reporting.create_report(report)
+
+
+def check_mismatched_kernel_versions():
+    # Get running kernel version
+    running_kernel_version = _get_booted_kernel()
+
+    # Get default kernel version
+    # NOTE: grubby dependency?
+    try:
+        default_kernel_version = run(["grubby", "--default-kernel"])["stdout"].strip()
+        # Remove the /boot/vmlinuz- prefix from the path
+        default_kernel_version = os.path.basename(default_kernel_version).split("-", 1)[1]
+    except (OSError, CalledProcessError):
+        api.current_logger().warning("Failed to query grubby for default kernel", exc_info=True)
+        return
+
+    if running_kernel_version != default_kernel_version:
+        summary = (
+            "This host is not running the default kernel version currently configured in the bootloader. "
+            "Maybe a new kernel version was installed without then rebooting the host? "
+            "Policy requires the host to be rebooted before going forward with the upgrade."
+        )
+
+        kernel_versions_related = [
+            reporting.RelatedResource("default_kernel_version", default_kernel_version),
+            reporting.RelatedResource("running_kernel_version", running_kernel_version),
+        ]
+
+        report = [
+            reporting.Title("Mismatched kernel versions detected"),
+            reporting.Summary(summary),
+            reporting.Severity(reporting.Severity.HIGH),
+            reporting.Groups([reporting.Groups.SANITY, reporting.Groups.INHIBITOR]),
+            reporting.Remediation(hint="Please reboot the host machine."),
+        ] + kernel_versions_related
+        reporting.create_report(report)
```

Comment:
@swapdisk What is exact purpose of this check? Is there a benefit to block the upgrade, when you are booted in a newer kernel than the default is set? (the last installed kernel is set always as default, even when an older version (in comparison to the currently booted kernel) is installed as the last one). Currently, the check, whether the system is booted in the newest installed kernel is present for many years. So curious what is the purpose of this one? Also, in case the default kernel is older than the newest one, when this actor is present, the upgrade is always inhibit, as this or our actor block the upgrade. 

The upstream actor: https://github.com/oamg/leapp-repository/tree/master/repos/system_upgrade/common/actors/kernel/checkinstalledkernels

Reply:
Right, I didn't know there was already a check for booted in the newest installed kernel present for many years. So this check for booted to default kernel version is a little redundant, but still useful as an example of how to make a custom Leapp actor to enforce an organization's policies.